lib/session-state.js

```diff
@@ -0,0 +1,593 @@
+const crypto = require('hypercore-crypto')
+const b4a = require('b4a')
+const assert = require('nanoassert')
+const flat = require('flat-tree')
+const quickbit = require('quickbit-universal')
+const Mutex = require('./mutex')
+const Bitfield = require('./bitfield')
+// const Info = require('./info')
+const { INVALID_OPERATION, INVALID_SIGNATURE } = require('hypercore-errors')
+
+module.exports = class SessionState {
+  constructor (core, storage, blocks, tree, snapshotLength, name) {
+    this.core = core
+    this.index = this.core.sessionStates.push(this) - 1
+
+    this.storage = storage
+    this.name = name
+    this.sessions = []
+
+    this.mutex = new Mutex()
+
+    this.blocks = blocks
+    this.tree = tree
+    this.snapshotLength = snapshotLength
+    this.snapshotCompatLength = snapshotLength
+
+    this.active = 0
+
+    this._onflush = null
+    this._flushing = null
+    this._activeBatch = null
+  }
+
+  isSnapshot () {
+    return this.storage.snapshotted
+  }
+
+  isDefault () {
+    return this.core.state === this
+  }
+
+  addSession (s) {
+    s._stateIndex = this.sessions.push(s) - 1
+  }
+
+  removeSession (s) {
+    if (s._stateIndex === -1) return
+    const head = this.sessions.pop()
+    if (head !== s) this.sessions[(head._stateIndex = s._stateIndex)] = head
+    s._stateIndex = -1
+  }
+
+  flushedLength () {
+    if (this.isDefault()) return this.tree.length
+    if (this.isSnapshot()) return this.snapshotLength
+
+    return this.storage.dependencyLength()
+  }
+
+  unref () {
+    if (--this.active > 0) return
+    this.destroy()
+  }
+
+  ref () {
+    this.active++
+    return this
+  }
+
+  destroy () {
+    if (this.index === -1) return
+
+    this.active = 0
+    this.storage.destroy()
+    this.mutex.destroy(new Error('Closed')).catch(noop)
+
+    const head = this.core.sessionStates.pop()
+    if (head !== this) this.core.sessionStates[(head.index = this.index)] = head
+
+    this.index = -1
+  }
+
+  snapshot () {
+    const s = new SessionState(
+      this.core,
+      this.storage.snapshot(),
+      this.blocks,
+      this.tree, // todo: should clone also but too many changes atm
+      this.tree.length,
+      this.name
+    )
+
+    return s
+  }
+
+  memoryOverlay () {
+    const storage = this.storage.createMemoryOverlay()
+    const s = new SessionState(
+      this.core,
+      storage,
+      this.blocks,
+      this.tree.clone(storage),
+      -1,
+      this.name
+    )
+
+    return s
+  }
+
+  _updateDependencies (dependency) {
```

Comment:
mb make this public since its called from core? non of these are meant for public users anyway to call